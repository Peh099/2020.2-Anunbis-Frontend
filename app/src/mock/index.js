import MockAdapter from 'axios-mock-adapter';
import api from '../services/Api';


const mock = new MockAdapter(api);

export default mock