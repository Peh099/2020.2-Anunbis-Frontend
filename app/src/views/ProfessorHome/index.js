import React from 'react';
import Feed from '../../components/Feed';
import btnOptions from '../../assets/images/Btn_options.png';
import {
  Container,
  ContainerPost,
  ImageOptions,
  ContainerOptions,
  ContainerHeader,
  BtnHomeProfessor,
  Home,
  LoadingBox,
  DisciplinePostsStyle,
} from './styles';
import Button from '../../components/Button';
import Loading from '../../components/Loading';
import { getPosts } from '../../services/Posts';
import Graphic from '../../components/Statistics';

const DisciplineContent = ({ discipline }) => {
  const [boxPost, setBoxPost] = React.useState(false);
  const [rotate, setRotate] = React.useState('');

  function handleSetBoxPost() {
    setRotate(!boxPost ? '90deg' : '');
    setBoxPost(!boxPost);
  }

  return (
    <ContainerPost>
      <DisciplineOptions onClick={handleSetBoxPost}>
        <ImageOptions src={btnOptions} rotate={rotate} />
        {`[${discipline.discipline_code}] ${discipline.name}`}
      </DisciplineOptions>
      {boxPost && <DisciplinePosts discipline={discipline} />}
    </ContainerPost>
  );
};

const DisciplineOptions = ({ children, onClick }) => (
  <ContainerOptions onClick={() => onClick()}>{children}</ContainerOptions>
);

const DisciplinePosts = ({ discipline }) => (
  <DisciplinePostsStyle>
    <ContainerHeader>
      <Feed.Header
        professor={discipline}
        feedbacks={discipline.feedbacks}
        canAvaliate={false}
      />
    </ContainerHeader>
    <Feed.PostsBox posts={discipline.posts} canReport />
  </DisciplinePostsStyle>
);

function sortPostsByDiscipline(posts) {
  const disciplines = {};
  const arrayFeedbacks = {};
  if (Object.keys(posts).length !== 0) {
    posts.forEach((postKey) => {
      if (disciplines[postKey.discipline.discipline_code] === undefined) {
        disciplines[postKey.discipline.discipline_code] = {
          feedbacks: {},
          posts: [],
          discipline_code: postKey.discipline.discipline_code,
          name: postKey.discipline.name,
        };
      }
      disciplines[postKey.discipline.discipline_code].posts.push(postKey);
    });
    const accumulatorInicial = {
      rating: 0,
      didactic: 0,
      metod: 0,
      avaliations: 0,
      disponibility: 0,
    };
    Object.keys(disciplines).forEach((disciplineCode) => {
      arrayFeedbacks[disciplineCode] = disciplines[disciplineCode].posts.reduce(
        (acumulador, valorAtual) => ({
          rating: acumulador.rating + valorAtual.rating,
          didactic: acumulador.didactic + valorAtual.didactic,
          metod: acumulador.metod + valorAtual.metod,
          avaliations: acumulador.avaliations + valorAtual.avaliations,
          disponibility: acumulador.disponibility + valorAtual.disponibility,
        }),
        accumulatorInicial,
      );
      disciplines[disciplineCode].feedbacks = {
        rating:
          arrayFeedbacks[disciplineCode].rating /
          disciplines[disciplineCode].posts.length,
        didactic:
          arrayFeedbacks[disciplineCode].didactic /
          disciplines[disciplineCode].posts.length,
        metod:
          arrayFeedbacks[disciplineCode].metod /
          disciplines[disciplineCode].posts.length,
        avaliations:
          arrayFeedbacks[disciplineCode].avaliations /
          disciplines[disciplineCode].posts.length,
        disponibility:
          arrayFeedbacks[disciplineCode].disponibility /
          disciplines[disciplineCode].posts.length,
      };
    });
  }
  return disciplines;
}

export default function ProfessorHome() {
  const [isStatistics, setIsStatistics] = React.useState(false);
  const [disciplines, setDisciplines] = React.useState({});
  const [loading, setLoading] = React.useState(true);

  React.useEffect(() => {
    getPosts((post) => {
      setDisciplines(sortPostsByDiscipline(post));
      setTimeout(() => setLoading(false), 500);
    });
  }, []);
  function OptionsProfessorHome() {
    return (
      <BtnHomeProfessor>
        <Button
          text="AVALIAÇÕES"
          backColor="var(--yellow)"
          padding="15px 40px"
          radius="20px"
          onClick={() => setIsStatistics(false)}
        />
        <Button
          text="ESTATÍSTICA"
          backColor="var(--yellow)"
          padding="15px 40px"
          radius="20px"
          onClick={() => setIsStatistics(true)}
        />
      </BtnHomeProfessor>
    );
  }

  return (
    <Home>
      <OptionsProfessorHome />
      {isStatistics && (
        <Feed title="Estatística">
          {Object.keys(disciplines).length === 0 && !loading ? (
            <Container>Sem avaliações ainda</Container>
          ) : (
            <Graphic />
          )}
        </Feed>
      )}
      {!isStatistics && (
        <Feed title="Avaliações sobre você">
          {Object.keys(disciplines).length === 0 && !loading ? (
            <Container>Sem avaliações ainda</Container>
          ) : (
            <Container backColor="#FFD54F">
              {!loading &&
                Object.keys(disciplines).map((dis) => (
                  <DisciplineContent key={dis} discipline={disciplines[dis]} />
                ))}
              {loading && (
                <LoadingBox>
                  <Loading />
                </LoadingBox>
              )}
            </Container>
          )}
        </Feed>
      )}
    </Home>
  );
}
