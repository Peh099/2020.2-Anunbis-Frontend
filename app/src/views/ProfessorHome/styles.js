import styled, { keyframes } from 'styled-components';

const translate = keyframes`
    from{
        opacity: 0.7;
        transform: translateY(10%);
    }
    to{
        opacity: 1;
        transform: translateY(0%);
    }
`;

export const Container = styled.div`
  border-radius: 20px;
  background: ${(props) => (props.backColor ? props.backColor : '')};
  min-height: 180px;
  height: fit-content;
  box-shadow: ${(props) =>
    props.backColor ? '2px 2px 2px rgba(0, 0, 0, 50%)' : 0};
  display: flex;
  align-items: center;
  flex-direction: column;
  margin: 10px;
  padding-bottom: 20px;
  overflow-y: auto;
`;

export const ContainerPost = styled.div`
  background-color: #fff9c4;
  box-shadow: 2px 2px 2px rgba(0, 0, 0, 50%);
  margin-top: 20px;
  border-radius: 15px;
  height: fit-content;
  width: 97%;
  animation: ${translate} 1s ease;
`;

export const DisciplinePostsStyle = styled.div`
  animation: ${translate} 1.5s ease;
`;

export const ContainerOptions = styled.div`
  display: grid;
  grid-template-columns: 100px 400px;
  align-items: center;
  justify-content: space-evenly;
  font-weight: bold;
  padding: 5px;
`;

export const ImageOptions = styled.img`
  height: 45px;
  transition: transform 1s ease;
  transform: ${(props) => (props.rotate ? `rotate(${props.rotate})` : '')};
`;

export const ContainerHeader = styled.div`
  display: flex;
  padding-left: 60px;
`;

export const BtnHomeProfessor = styled.div`
  background-color: var(--lightWhite);
  margin: 90px 100px 0 0;
  height: 150px;
  width: 250px;
  border-radius: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
`;

export const Home = styled.div`
  display: grid;
  grid-template-columns: 300px 800px;
`;

export const LoadingBox = styled.div`
  width: 100%;
  height: 20vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;
