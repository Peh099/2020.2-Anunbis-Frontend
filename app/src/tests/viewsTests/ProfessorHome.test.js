import renderer from 'react-test-renderer';
import mock from '../../mock';
import { validPost } from '../../mock/fixtures/stored_post';
import ProfessorHome from '../../views/ProfessorHome';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';


mock.onGet('post').reply(200,
  [validPost]
)

describe('Snapshot ProfessorHome view', () => {
  it('Assyncronous test for useEffect', async () => {
    render(<ProfessorHome />)

    const loadBox = screen.getByTestId('loading')

    await waitFor(() => {
      expect(loadBox).toBeInTheDocument()
    })
    await waitFor(() => {
      expect(screen.getByText(`[${validPost.discipline.discipline_code}] ${validPost.discipline.name}`)).toBeInTheDocument()
    })
    await waitFor(() => {
      expect(loadBox).not.toBeInTheDocument();
    })
   })
  it('Clicking on buttons and expecting changes', async () => {
    render(<ProfessorHome />)
    
    expect(screen.getByText('Avaliações sobre você')).toBeTruthy()

    fireEvent.click(screen.getByText('ESTATÍSTICAS'))

    await waitFor(() => {
      expect(screen.getByText('Estatísticas')).toBeTruthy()
    })

    fireEvent.click(screen.getByText('AVALIAÇÕES'))

    await waitFor(() => {
      expect(screen.getByText('Avaliações sobre você')).toBeTruthy()
    })

  })
 });