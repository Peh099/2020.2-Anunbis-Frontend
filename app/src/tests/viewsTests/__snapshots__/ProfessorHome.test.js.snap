// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot ProfessorHome view matches the snapshot 1`] = `
<div
  className="sc-carFqZ ihJGFH"
>
  <div
    className="sc-jcwpoC fmcQXe"
  >
    <button
      className="sc-iCoGMd bDcoSI"
      onClick={[Function]}
      radius="20px"
    >
      AVALIAÇÕES
    </button>
    <button
      className="sc-iCoGMd bDcoSI"
      onClick={[Function]}
      radius="20px"
    >
      ESTATÍSTICAS
    </button>
  </div>
  <div
    className="sc-bdnxRM dHvxIB"
    title="Avaliações sobre você"
  >
    <div
      className="sc-gtsrHT iUwlZJ"
    >
      Avaliações sobre você
    </div>
    <div
      className="sc-iklJeh czotxj"
    >
      <div
        className="sc-iTVJFM cWVcHr"
      >
        <div
          className="sc-efHYUO kREmPd"
        >
          <span
            className="sc-iBzEeX cfOQfB"
          />
          <span
            className="sc-iBzEeX kcSDut"
          />
          <span
            className="sc-iBzEeX kCIfxu"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
